### ISSUE TITLE
Performance degradation when resolving cookbooks on chef-client 14

### ISSUE THREAD

**Issue by jwadolowski** [chef/chef#7230]
## Description

Over the last couple of weeks I've been testing chef-client v14.0.202 and noticed significant difference when it comes to execution times in comparison to v13.8.5.

Exactly the same run list (custom set of cookbooks) took either 13s (chef 13) or 1m 32s (chef 14). All resources have been already provisioned, so it's definitely not a side-effect of initial setup. The slow-down was observed right after the following line in the output

```
resolving cookbooks for run list: ["myproject-httpd::_hosts", "myproject-httpd::dispatcher"]
```

To make sure it's not caused by my cookbooks I replicated the same behaviour using [java](https://github.com/sous-chefs/java) cookbook (see details outlined below).

## Chef Version

* 14.0.202
* 13.8.5

## Platform Version

```
$ cat /etc/redhat-release
CentOS Linux release 7.4.1708 (Core)
```

## Replication Case

The entire procedure was observed on local Vagrant instance using [Chef Solo Provisioner](https://www.vagrantup.com/docs/provisioning/chef_solo.html), however to minimize the impact of the Vagrant itself I run `chef-solo` command by hand as presented below.

All cookbooks are available in `/tmp/vagrant-chef/1b6f8924afde185428242d2af14667fc/cookbooks` directory. 

Client configuration:

```
$ cat /tmp/vagrant-chef/solo.rb
node_name "myproject-vagrant"
file_cache_path    "/var/chef/cache"
file_backup_path   "/var/chef/backup"
cookbook_path ["/tmp/vagrant-chef/1b6f8924afde185428242d2af14667fc/cookbooks"]
log_level :info
verbose_logging false

encrypted_data_bag_secret nil



environment_path "/tmp/vagrant-chef/5fc8dd935f5215fbfb00e5437e3fcb4c/environments"

environment "myproject-vagrant"


http_proxy nil
http_proxy_user nil
http_proxy_pass nil
https_proxy nil
https_proxy_user nil
https_proxy_pass nil
no_proxy nil
```

Two run-lists were used:

```
$ cat /tmp/vagrant-chef/dna.json
{
  "run_list": [
    "recipe[myproject-httpd::_hosts]",
    "recipe[myproject-httpd::dispatcher]"
  ]
}
```

and 

```
$ cat /tmp/vagrant-chef/dna.json
{
  "run_list": [
    "recipe[java::default]"
  ]
}
```

### Custom cookbooks (chef 14)

```
$ sudo /opt/chef/embedded/bin/ruby --disable-gems /bin/chef-solo --config /tmp/vagrant-chef/solo.rb --json-attributes /tmp/vagrant-chef/dna.json --log_level info --no-color --force-formatter
[2018-05-07T21:35:43+01:00] INFO: Started chef-zero at chefzero://localhost:1 with repository at /tmp/vagrant-chef/1b6f8924afde185428242d2af14667fc
  One version per cookbook
  environments at /tmp/vagrant-chef/5fc8dd935f5215fbfb00e5437e3fcb4c/environments

Starting Chef Client, version 14.0.202
[2018-05-07T21:35:43+01:00] INFO: *** Chef 14.0.202 ***
[2018-05-07T21:35:43+01:00] INFO: Platform: x86_64-linux
[2018-05-07T21:35:43+01:00] INFO: Chef-client pid: 12434
[2018-05-07T21:35:43+01:00] INFO: The plugin path /etc/chef/ohai/plugins does not exist. Skipping...
[2018-05-07T21:35:46+01:00] INFO: Setting the run_list to ["recipe[myproject-httpd::_hosts]", "recipe[myproject-httpd::dispatcher]"] from CLI options
[2018-05-07T21:35:46+01:00] INFO: Run List is [recipe[myproject-httpd::_hosts], recipe[myproject-httpd::dispatcher]]
[2018-05-07T21:35:46+01:00] INFO: Run List expands to [myproject-httpd::_hosts, myproject-httpd::dispatcher]
[2018-05-07T21:35:46+01:00] INFO: Starting Chef Run for myproject-vagrant
[2018-05-07T21:35:46+01:00] INFO: Running start handlers
[2018-05-07T21:35:46+01:00] INFO: Start handlers complete.
resolving cookbooks for run list: ["myproject-httpd::_hosts", "myproject-httpd::dispatcher"]

---
>>>
>>>
>>> This is where it got stuck for around a minute
>>>
>>>
---

[2018-05-07T21:37:06+01:00] INFO: Loading cookbooks [myproject-httpd@0.1.4, cognifide-httpd@2.0.15, hostsfile@3.0.1, git@9.0.0, logrotate@2.2.0, build-essential@8.1.1, seven_zip@2.0.2, mingw@2.0.2, windows@4.2.2]
...
[2018-05-07T21:37:08+01:00] INFO: Chef Run complete in 82.719818151 seconds

Running handlers:
[2018-05-07T21:37:08+01:00] INFO: Running report handlers
Running handlers complete
[2018-05-07T21:37:08+01:00] INFO: Report handlers complete
Chef Client finished, 0/30 resources updated in 01 minutes 24 seconds
```

### Custom cookbooks (chef 13)

```
$ sudo /opt/chef/embedded/bin/ruby --disable-gems /bin/chef-solo --config /tmp/vagrant-chef/solo.rb --json-attributes /tmp/vagrant-chef/dna.json --log_level info --no-color --force-formatter
[2018-05-07T21:46:55+01:00] INFO: Started chef-zero at chefzero://localhost:1 with repository at /tmp/vagrant-chef/1b6f8924afde185428242d2af14667fc
  One version per cookbook
  environments at /tmp/vagrant-chef/5fc8dd935f5215fbfb00e5437e3fcb4c/environments

[2018-05-07T21:46:55+01:00] INFO: Forking chef instance to converge...
Starting Chef Client, version 13.8.5
[2018-05-07T21:46:55+01:00] INFO: *** Chef 13.8.5 ***
[2018-05-07T21:46:55+01:00] INFO: Platform: x86_64-linux
[2018-05-07T21:46:55+01:00] INFO: Chef-client pid: 15520
[2018-05-07T21:46:55+01:00] INFO: The plugin path /etc/chef/ohai/plugins does not exist. Skipping...
[2018-05-07T21:46:57+01:00] INFO: Setting the run_list to ["recipe[myproject-httpd::_hosts]", "recipe[myproject-httpd::dispatcher]"] from CLI options
[2018-05-07T21:46:57+01:00] INFO: Run List is [recipe[myproject-httpd::_hosts], recipe[myproject-httpd::dispatcher]]
[2018-05-07T21:46:57+01:00] INFO: Run List expands to [myproject-httpd::_hosts, myproject-httpd::dispatcher]
[2018-05-07T21:46:57+01:00] INFO: Starting Chef Run for myproject-vagrant
[2018-05-07T21:46:57+01:00] INFO: Running start handlers
[2018-05-07T21:46:57+01:00] INFO: Start handlers complete.
resolving cookbooks for run list: ["myproject-httpd::_hosts", "myproject-httpd::dispatcher"]

---
>>>
>>>
>>> It stops here for a short period, but the pause is much shorter than in previous case 
>>>
>>>
---

[2018-05-07T21:47:03+01:00] INFO: Loading cookbooks [myproject-httpd@0.1.4, cognifide-httpd@2.0.15, hostsfile@3.0.1, git@9.0.0, logrotate@2.2.0, build-essential@8.1.1, seven_zip@2.0.2, mingw@2.0.2, windows@4.2.2]
...
[2018-05-07T21:47:09+01:00] INFO: Chef Run complete in 11.99977357 seconds

Running handlers:
[2018-05-07T21:47:09+01:00] INFO: Running report handlers
Running handlers complete
[2018-05-07T21:47:09+01:00] INFO: Report handlers complete
Chef Client finished, 0/30 resources updated in 14 seconds
```

### Community cookbook (chef 14)

```
$ sudo /opt/chef/embedded/bin/ruby --disable-gems /bin/chef-solo --config /tmp/vagrant-chef/solo.rb --json-attributes /tmp/vagrant-chef/dna.json --log_level info --no-color --force-formatter
[2018-05-07T21:55:09+01:00] INFO: Started chef-zero at chefzero://localhost:1 with repository at /tmp/vagrant-chef/1b6f8924afde185428242d2af14667fc
  One version per cookbook
  environments at /tmp/vagrant-chef/5fc8dd935f5215fbfb00e5437e3fcb4c/environments

Starting Chef Client, version 14.0.202
...
[2018-05-07T21:55:44+01:00] INFO: Chef Run complete in 32.231628317 seconds

Running handlers:
[2018-05-07T21:55:44+01:00] INFO: Running report handlers
Running handlers complete
[2018-05-07T21:55:44+01:00] INFO: Report handlers complete
Chef Client finished, 0/5 resources updated in 34 seconds
```

### Community cookbook (chef 13)

```
$ sudo /opt/chef/embedded/bin/ruby --disable-gems /bin/chef-solo --config /tmp/vagrant-chef/solo.rb --json-attributes /tmp/vagrant-chef/dna.json --log_level info --no-color --force-formatter
[2018-05-07T21:57:26+01:00] INFO: Started chef-zero at chefzero://localhost:1 with repository at /tmp/vagrant-chef/1b6f8924afde185428242d2af14667fc
  One version per cookbook
  environments at /tmp/vagrant-chef/5fc8dd935f5215fbfb00e5437e3fcb4c/environments

[2018-05-07T21:57:26+01:00] INFO: Forking chef instance to converge...
Starting Chef Client, version 13.8.5
...
[2018-05-07T21:57:42+01:00] INFO: Chef Run complete in 13.303452913 seconds

Running handlers:
[2018-05-07T21:57:42+01:00] INFO: Running report handlers
Running handlers complete
[2018-05-07T21:57:42+01:00] INFO: Report handlers complete
Chef Client finished, 0/5 resources updated in 15 seconds
```

## Stacktrace

N/A

**Comment by thommay** [comment]
I can't reproduce your results; using the following Vagrantfile:

```
Vagrant.configure("2") do |config|
  config.vm.box = "bento/centos-7"
  config.vm.provision "chef_solo" do |chef|
    chef.version = "14.1.1"
    chef.add_recipe "java::default"
  end
end
```

And trying the following versions:
 * 14.1.1 : 3.2  seconds
 * 14.0.202 : 3.7 seconds
 * 13.9.1 : 4.1 seconds

We'll need to see the output using `-l debug` in 14 to see where the time is being taken.

**Comment by jwadolowski** [comment]
Hi @thommay. 

Here's the debug output from `14.0.202`: https://gist.github.com/jwadolowski/3bf9cf314441553b1592a7fcb5cdf2fd (I removed a few lines for brevity, can upload the whole thing if needed).

The moment it stops for noticeable amount of time:

```
...
resolving cookbooks for run list: ["java::default"]
[2018-05-10T17:18:32+01:00] DEBUG: POST /organizations/chef/environments/proj-vagrant/cookbook_versions
...
[2018-05-10T17:18:54+01:00] DEBUG:
--- RESPONSE (200) ---
...
```

As far as I understand `POST /organizations/chef/environments/proj-vagrant/cookbook_versions` was sent at 17:18:32, but response got back at 17:18:54 (took 22 seconds to process that).

**Comment by jwadolowski** [comment]
Same output for `13.8.5`: https://gist.github.com/jwadolowski/c51ee0c630cdefb967322c96797b86af

`POST /organizations/chef/environments/proj-vagrant/cookbook_versions` request took only 4 seconds (started at 17:37:52, completed at 17:37:56)

**Comment by lamont-granquist** [comment]
this is fixed in 14.1.1?

**Comment by lamont-granquist** [comment]
should have been fixed in 14.0.210 via https://github.com/chef/chef/pull/7167

**Comment by jwadolowski** [comment]
Yep, I've just tested the same scenario on 14.1.1/14.1.12 and it is much faster.
